lib/OpenQA/Schema/Result/JobGroups.pm

```diff
@@ -284,4 +288,193 @@ sub tags {
+        my $yaml_defaults_for_arch = $yaml_defaults->{$arch};
+        foreach my $product_name (sort keys %$yaml_products_for_arch) {
+            foreach my $spec (@{$yaml_products_for_arch->{$product_name}}) {
+                # Get testsuite, machine, prio and job template settings from YAML data
+                my $testsuite_name;
+                my $job_template_name;
+                # Assign defaults
+                my $prio          = $yaml_defaults_for_arch->{priority};
+                my $machine_names = $yaml_defaults_for_arch->{machine};
+                my $settings      = dclone($yaml_defaults_for_arch->{settings} // {});
+                if (ref $spec eq 'HASH') {
+                    # We only have one key. Asserted by schema
+                    $testsuite_name = (keys %$spec)[0];
+                    my $attr = $spec->{$testsuite_name};
+                    if ($attr->{priority}) {
+                        $prio = $attr->{priority};
+                    }
+                    if ($attr->{machine}) {
+                        $machine_names = $attr->{machine};
+                    }
+                    if ($attr->{testsuite}) {
+                        $job_template_name = $testsuite_name;
+                        $testsuite_name    = $attr->{testsuite};
+                    }
+                    if ($attr->{settings}) {
+                        %$settings = (%{$settings // {}}, %{$attr->{settings}});
+                    }
+                }
+                else {
+                    $testsuite_name = $spec;
+                }
+
+                $machine_names = [$machine_names] if ref($machine_names) ne 'ARRAY';
+                foreach my $machine_name (@{$machine_names}) {
+                    my $job_template_key
+                      = $arch . $product_name . $machine_name . $testsuite_name . ($job_template_name // '');
+                    die "Job template name '"
+                      . ($job_template_name // $testsuite_name)
+                      . "' is defined more than once. "
+                      . "Use a unique name and specify 'testsuite' to re-use test suites in multiple scenarios.\n"
+                      if $job_template_names{$job_template_key};
+                    $job_template_names{$job_template_key} = {
+                        prio              => $prio,
+                        machine_name      => $machine_name,
+                        arch              => $arch,
+                        product_name      => $product_name,
+                        product_spec      => $yaml_products->{$product_name},
+                        job_template_name => $job_template_name,
+                        testsuite_name    => $testsuite_name,
+                        settings          => $settings
+                    };
+                }
+            }
+        }
+    }
+
+    return \%job_template_names;
+}
+
+sub expand_yaml {
+    my ($self, $job_template_names) = @_;
```

Comment:
It strikes me as odd that this is sub takes `$self` but doesn't use it. The result is not the expansion of what's in the group, but the expansion of the given hash. :thinking:

Reply:
I actually thought about putting some functions into their own module because they do not depend on database objects, but skipped it for now.